### PR TITLE
New Adapter: Admatic

### PR DESCRIPTION
**Pull request by bakicam**
We added s2s support for Admatic.

Docs PR: https://github.com/prebid/prebid.github.io/pull/5232

Old Pull Request reference: https://github.com/prebid/prebid-server/pull/3608

### CODE REVIEW

**Comment by SyntaxNode** [comment]
The validation tests are failing because you need to format your code `go fmt`.

**Comment by bakicam** [comment]
> The validation tests are failing because you need to format your code `go fmt`.
@SyntaxNode 
Hello,
go fmt was done. Can you check?

**Comment by bakicam** [comment]
Hi,
We need for your support.

**Comment by onkarvhanumante** [comment]
> Hi, We need for your support.
@bakicam  to merge `Prebid:master` changes into `Admatic-Sofware:master` and update MakeRequests to return ImpIDs as mentioned below

https://github.com/prebid/prebid-server/blob/763551963dd3208dde6f9bf35b31ef76cc83814f/adapters/pubmatic/pubmatic.go#L175-L182

**Comment by bakicam** [comment]
Hello @onkarvhanumante , 
We made the solution.

**Comment by farukcam** [comment]
Hi @onkarvhanumante,
We need for your support.

**Comment by onkarvhanumante** [comment]
> Hi @onkarvhanumante, We need for your support.

Refer https://github.com/prebid/prebid-server/pull/3654#issuecomment-2106983608 to fix PR tests. let me know if any other support is needed


```
--- FAIL: TestJsonSamples (0.00s)
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:1[28](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:29)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:5[35](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:36)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/exemplary/banner.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:[41](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:42)2
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:[44](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:45)5
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/exemplary/native.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:4[45](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:46)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:[46](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:47)7
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/exemplary/optional-params.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:[53](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:54)5
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/exemplary/video.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:2[56](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:57)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/supplemental/response-200-without-body.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:76
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:4[67](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:68)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:[70](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:71)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/supplemental/response-204.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
    test_json.go:128: 
        	Error Trace:	/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:256
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:412
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:128
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:99
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:[76](https://github.com/prebid/prebid-server/actions/runs/9075060288/job/24934940735?pr=3654#step:5:77)
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:445
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:467
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/path.go:535
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/test_json.go:70
        	            				/home/runner/work/prebid-server/prebid-server/adapters/admatic/admatic_test.go:20
        	Error:      	Received unexpected error:
        	            	expected.ImpIDs must contain at least one imp ID
        	Test:       	TestJsonSamples
        	Messages:   	admatictest/supplemental/server-error.json Expected RequestData was not returned by adapters' MakeRequests() implementation: httpRequest[0]
```

**Comment by bakicam** [comment]
Hi @onkarvhanumante,
We fixed all problems. We need for your support.

File: adapters/admatic/admatic.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo
+			} else if imp.Native != nil {
+				return openrtb_ext.BidTypeNative
+			}
+		}
+	}
+	return ""
```
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:147]: @bakicam consider returning error when no bid type found

Comment (bakicam) [reply adapters/admatic/admatic.go]: Hi @onkarvhanumante,
Problem solved.

**Comment by bakicam** [comment]
Hi @onkarvhanumante,
Can you check again?

File: adapters/admatic/admatic.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func getMediaTypeForBid(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner != nil {
+				return openrtb_ext.BidTypeBanner, nil
+			} else if imp.Video != nil {
+				return openrtb_ext.BidTypeVideo, nil
+			} else if imp.Native != nil {
```
[the remark this answers is not in the document]
Comment (onkarvhanumante) [reply adapters/admatic/admatic.go:147]: @bakicam  any feedback on above ^^ ?

[the remark this answers is not in the document]
Comment (farukcam) [reply adapters/admatic/admatic.go:147]: Hello @onkarvhanumante, 
We will improve MType support later.

**Comment by farukcam** [comment]
Hello @onkarvhanumante 
I updated the fork repo. Is it ready to merge?

**Comment by farukcam** [comment]
Hi @onkarvhanumante,
We need for your support. Thank you so much!

**Review by onkarvhanumante — APPROVED** [review]

**Review by farukcam — APPROVED** [review]

**Comment by farukcam** [comment]
@Sonali-More-Xandr @onkarvhanumante Hello,
I request your support for the merge. Thank you very much.

**Comment by farukcam** [comment]
@Sonali-More-Xandr @onkarvhanumante Hello,
I request your support for the merge. Thank you very much.

**Comment by farukcam** [comment]
@Sonali-More-Xandr @onkarvhanumante Hello,
I request your support for the merge. Thank you very much.

File: adapters/admatic/admatic.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	if err := json.Unmarshal(bidderExt.Bidder, &admaticExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &admaticExt, nil
+}
+
+// "Un-templates" the endpoint by replacing macroses and adding the required query parameters
```
Comment (SyntaxNode) [inline adapters/admatic/admatic.go:92]: Nitpick: Please follow Go comment convention by starting with the method name, something like:

```
// resolveUrl "un-templates" the endpoint by replacing macroses and adding the required query parameters
```

File: static/bidder-info/admatic.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+endpoint: "http://pbs.admatic.com.tr?host={{.Host}}"
+maintainer:
+  email: "prebid@admatic.com.tr"
+gvlVendorID: 1281
```
Comment (SyntaxNode) [inline static/bidder-info/admatic.yaml:4]: Confirmed GVL id matches with Admatic.

File: adapters/admatic/admatic.go (hunk 3 above)
Comment (bakicam) [reply adapters/admatic/admatic.go]: Hi,
This fixed

File: static/bidder-info/admatic.yaml (hunk 4 above)
Comment (bakicam) [reply static/bidder-info/admatic.yaml:4]: Hi,
It is right.

**Comment by farukcam** [comment]
Hello @onkarvhanumante , @Sonali-More-Xandr ,
Is there anything we should do?

**Review by onkarvhanumante — DISMISSED** [review]

**Comment by onkarvhanumante** [comment]
> Hello @onkarvhanumante , @Sonali-More-Xandr , Is there anything we should do?

@bakicam  thank you for keeping patience with PR review. PR needs one more approval. Will try to get PR reviewed for the remaining approval

File: adapters/admatic/admatictest/supplemental/server-error.json
Comment (bsardo) [inline adapters/admatic/admatictest/supplemental/server-error.json:1]: This test is the same as `response-204.json`. I think you meant to test a 4XX or 5XX http status code here. I see though that your `MakeBids` function does not handle these cases. I added a suggestion in `MakeBids` about how you can use a utility function to do so.

File: adapters/admatic/admatic.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		return "", err
+	}
+
+	return parsedUrl.String(), nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
```
Comment (bsardo) [inline adapters/admatic/admatic.go:112]: I suggest using the adapter utility function instead:
```
if adapters.IsResponseStatusCodeNoContent(responseData) {
	return nil, nil
}
```

File: adapters/admatic/admatic.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
```
Comment (bsardo) [inline adapters/admatic/admatic.go:113]: I see that you're not handling non-2XX HTTP status codes. You can handle them with an adapter utility function:
```
if err := adapters.CheckResponseStatusCodeForErrors(responseData); err != nil {
	return nil, []error{err}
}
```

File: static/bidder-params/admatic.json
Comment (bsardo) [inline static/bidder-params/admatic.json:1]: Your docs say that `networkId` is also required but I don't see it here or any reference to it in the adapter.

**Comment by bakicam** [comment]
Hi,
We fixed all problems. Can you check again?

File: adapters/admatic/admatic.go [hunk 7]
```diff
@@ -32,85 +31,67 @@ func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server co
 }
 
 func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var requests []*adapters.RequestData
+	var errs []error
 
-	host := ""
-
+	requestCopy := *request
 	for _, imp := range request.Imp {
```
Comment (bsardo) [inline adapters/admatic/admatic.go:38]: Given that this loop now creates a separate request for each imp, please either update an exemplary test case to test the multiple imp case or create a supplemental test case to cover it.

Additionally, if it is easy, I suggest adding a supplemental multi-imp test case where one imp has an error and the other does not. I think you can easily force an error when building the endpoint for a given imp by specifying an invalid type like an empty object for `host`.

Comment (bakicam) [reply adapters/admatic/admatic.go:38]: Hello @bsardo,
It returns an error when there is an invalid host or networkId in an imp array.

Example: 
Invalid request: request.imp[1].ext.prebid.bidder.admatic failed validation.
host: Invalid type. Expected: string, given: integer

**Comment by bakicam** [comment]
Hi,
We fixed problems. Can you check again?

File: adapters/admatic/admatic.go (hunk 7 above)
Comment (bsardo) [reply adapters/admatic/admatic.go:38]: Hi @bakicam, PBS core will perform bidder param validation upstream. However, I'm suggesting that you first add an exemplary test case for when there are multiple well-formed imps, and then a second supplemental one that demonstrates how your adapter works when there are multiple imps where one of the imps is malformed. You don't have any multiple imp tests at the moment.
If your adapter is given a request with multiple imps and one of the imps has an error, your adapter appears to discard the malformed imp while making a bid request for the well-formed imp.

Given that adapter JSON tests are scoped to the adapter code, bidder param validation should be bypassed. For the multiple imp with one malformed case, the following test can be added:
[multiple-imps-with-error.json](https://github.com/user-attachments/files/16069543/multiple-imps-with-error.json)

Comment (bsardo) [reply adapters/admatic/admatic.go:38]: @bakicam the new file looks good. As I mentioned, you might also want to add `exemplary/multiple-imps.json` that does not involve any errors and demonstrates that your adapter sends two separate requests, one for each imp, to your adapter. Last suggestion from me 😄

Comment (bakicam) [reply adapters/admatic/admatic.go:38]: Hello @bsardo , thank you very much for your valuable suggestions and help. I also added this file. 😊

**Review by bsardo — APPROVED** [review]

**Comment by bakicam** [comment]
Hi,
Do we need to do anything for it to merge?

**Review by onkarvhanumante — APPROVED** [review]

**Pull request merged**